src/goto-programs/remove_function_pointers.cpp

```diff
@@ -140,6 +162,317 @@ symbolt &remove_function_pointerst::new_tmp_symbol()
+  if(expr.id()==ID_index)
+  {
+    index_exprt index_expr=to_index_expr(expr);
+
+    array_exprt array_expr;
+    if(try_get_array_from_index(index_expr, array_expr))
+    {
+      mp_integer value;
+      if(try_evaluate_index_value(index_expr.index(), value))
+      {
+        const exprt &func_expr=array_expr.operands()[integer2size_t(value)];
+        exprt precise_match;
+        if(try_get_precise_call(func_expr, precise_match))
+        {
+          out_functions.push_back(precise_match);
+        }
+        else if(try_get_from_address_of(func_expr, out_functions))
+        {
+        }
+        else if(try_get_call_from_symbol(func_expr, out_functions))
+        {
+        }
+        else
+        {
+          return false;
+        }
+      }
+      else
+      {
+        // We don't know what index it is,
+        // but we know the value is from the array
+        for(const exprt &op : array_expr.operands())
+        {
+          exprt precise_match;
+          if(try_get_precise_call(op, precise_match))
+          {
+            out_functions.push_back(precise_match);
+          }
+          else if(try_get_from_address_of(op, out_functions))
+          {
+          }
+          else if(try_get_call_from_symbol(op, out_functions))
+          {
+          }
+          else
+          {
+            // return false?
+            // in this case there is an element
+          }
+        }
+        return out_functions.size() > 0;
+      }
+    }
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/*******************************************************************\
+
+Function: remove_function_pointerst::try_evaluate_index_value
+
+  Inputs:
+   index_expr - The expression of the index of the index expression (e.g.
+                index_exprt::index())
+
+ Outputs: Returns true if was able to find a constant value for the index
+          expression. If true, then out_array_index will be the index within
+          the array that the function pointer is pointing to.
+
+ Purpose: Given an index into an array, resolve, if possible, the index
+          that is being accessed. This deals with symbols and typecasts to
+          constant values.
+
+\*******************************************************************/
+
+bool remove_function_pointerst::try_evaluate_index_value(
+  const exprt &index_value_expr, mp_integer &out_array_index)
+{
+  if(index_value_expr.id()==ID_constant)
+  {
+    constant_exprt constant_expr=to_constant_expr(index_value_expr);
+    mp_integer array_index;
+    bool errors=to_integer(constant_expr, array_index);
+    if(!errors)
+    {
+      out_array_index=array_index;
+    }
+    return !errors;
+  }
+  else if(index_value_expr.id()==ID_symbol)
+  {
+    // Resolve symbol and try again
+    const symbolt &index_symbol=
+      symbol_table.lookup(index_value_expr.get(ID_identifier));
+
+    return try_evaluate_index_value(index_symbol.value, out_array_index);
+  }
+  else if(index_value_expr.id()==ID_typecast)
```

Comment:
I don't really know what ID_typecast is doing here, but it shows up when accessing the array by index - it is straightforward to resolve but feels a bit hard coded...